src/basic/terminal-util.c

```diff
@@ -857,6 +857,36 @@ unsigned lines(void) {
         return cached_lines;
 }
 
+int terminal_set_size_fd(int fd, unsigned rows, unsigned cols) {
+        struct winsize ws;
+
+        if (rows == 0 && cols == 0)
+                return 0;
+
+        if (rows > USHRT_MAX)
+                rows = USHRT_MAX;
+        if (cols > USHRT_MAX)
+                cols = USHRT_MAX;
+
+        if (ioctl(fd, TIOCGWINSZ, &ws) < 0)
+                return log_debug_errno(errno, "TIOCSWINSZ ioctl for setting TTY size failed, ignoring: %m");
+
+        if (rows == 0)
+                rows = ws.ws_row;
+        if (cols == 0)
+                cols = ws.ws_col;
+
+        ws = (struct winsize) {
+                .ws_row = rows,
+                .ws_col = cols,
+        };
```

Comment:
why override the full structure you got from the kernel? wouldn't it make more sense to pass it back to the kernel with only the minimal stuff overriden? moreover as i read the kernel sources every call to TIOCSWINSZ will trigger SIGWINCH on listening processes, hence I think it would make sense to suppress any changes that are unnecessary. Hence maybe:

```c
if (ioctl(fd, TIOCGWINSZ, &ws) < 0) 
        return …;

if ((rows == UINT_MAX || rows == ws.ws_row) &&
    (cols == UINT_MAX || cols == ws.ws_col))
        return 0;

if (rows != UINT_MAX)
        ws.ws_rows = rows;
if (cols != UINT_MAX)
        ws.ws_cols = cols;

if (ioctl(fd, TIOCSWINSZ, &ws) < 0) 
         …
```

Reply:
I think if we call `TIOCSWINSZ` we have to set everything in winsize, otherwise we're passing garbage to the kernel potentially no?

I did add an extra check to not set the size if what we're going to configure is the same as what's already configured.

Reply:
Ah like that, now I understand. Modified PR to only modify what's needed instead of everything.